File: asid.go
package id

import (
	"crypto/sha256"
	"fmt"
	"strings"

	"github.com/google/uuid"
)

// GenerateJDASID generates a Job Description ASID
// Format: JD + random(2) + company(5) + random(2) + role(7) + random(2) + location(5) + random(4) + type(3)
// Returns an ID like "JD47ACMEX22ENGINEE33SFBAY5E7AJOB"
//
// The components provide semantic hints while maintaining uniqueness:
//   - company: Helps identify the organization
//   - role: The role title/type
//   - location: Geographic location
//   - type: Job type indicator (default: "JOB")
//
// This format enables:
//   - Quick visual identification of JD source
//   - Deduplication by comparing semantic components
//   - Versioning of the same JD over time (different random parts)
func GenerateJDASID(company, roleTitle, location string) (string, error) {
	// Generate a new UUID for the random components
	u, err := uuid.NewRandom()
	if err != nil {
		return "", fmt.Errorf("failed to generate UUID: %w", err)
	}

	// Get random hex characters from UUID (remove dashes)
	uuidStr := strings.ToUpper(strings.ReplaceAll(u.String(), "-", ""))

	// Extract random segments: 2 + 2 + 2 + 4 = 10 chars total
	rand1 := uuidStr[0:2]  // First 2 chars
	rand2 := uuidStr[2:4]  // Next 2 chars
	rand3 := uuidStr[4:6]  // Next 2 chars
	rand4 := uuidStr[6:10] // Next 4 chars

	// Extract vanity components
	companyPart := extractVanityComponent(company, 5)
	rolePart := extractVanityComponent(roleTitle, 7)
	locationPart := extractVanityComponent(location, 5)
	typePart := "JOB" // Default type indicator for job descriptions

	// Construct the JD ASID - 32 characters total
	// Format: JD(2) + random(2) + company(5) + random(2) + role(7) + random(2) + location(5) + random(4) + type(3) = 32
	jdID := fmt.Sprintf("JD%s%s%s%s%s%s%s%s",
		rand1,
		companyPart,
		rand2,
		rolePart,
		rand3,
		locationPart,
		rand4,
		typePart,
	)

	return jdID, nil
}

// GenerateJDASIDWithRetry generates a JD ASID with collision detection
// The checkExists function should return true if the ID already exists
func GenerateJDASIDWithRetry(company, roleTitle, location string, checkExists func(string) bool) (string, error) {
	const maxRetries = 10

	for i := 0; i < maxRetries; i++ {
		jdID, err := GenerateJDASID(company, roleTitle, location)
		if err != nil {
			return "", err
		}

		if checkExists != nil && checkExists(jdID) {
			continue // Collision detected, retry
		}

		return jdID, nil
	}

	return "", fmt.Errorf("failed to generate unique JD ASID after %d attempts", maxRetries)
}

// extractVanityComponent extracts the specified number of characters from a string, using hash-based padding
func extractVanityComponent(s string, length int) string {
	if len(s) == 0 {
		// For empty strings, generate from hash of empty string
		return generateHashBasedPadding("", length)
	}

	// Clean and uppercase the string, keeping only alphanumeric characters
	var cleaned strings.Builder
	for _, r := range strings.ToUpper(s) {
		if (r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9') {
			cleaned.WriteRune(r)
		}
	}

	cleanedStr := cleaned.String()
	if len(cleanedStr) == 0 {
		// If no valid characters, generate from hash of original string
		return generateHashBasedPadding(s, length)
	}

	// Take first N characters, pad with hash-based characters if needed
	if len(cleanedStr) >= length {
		return cleanedStr[:length]
	}

	// Need padding - use hash-based approach
	needed := length - len(cleanedStr)
	padding := generateHashBasedPadding(s, needed)
	return cleanedStr + padding
}

// generateHashBasedPadding generates deterministic padding characters from string hash
func generateHashBasedPadding(s string, length int) string {
	if length <= 0 {
		return ""
	}

	// Valid characters for vanity components (alphanumeric, excluding confusing ones)
	validChars := "ABCDEFGHJKLMNPQRSTUVWXYZ23456789" // No I,O,0,1 for clarity

	// Generate hash of the original string
	hash := sha256.Sum256([]byte(s))

	var result strings.Builder
	for i := 0; i < length; i++ {
		// Use different bytes from the hash for each position
		byteIndex := i % len(hash)
		charIndex := int(hash[byteIndex]) % len(validChars)
		result.WriteByte(validChars[charIndex])
	}

	return result.String()
}

// GenerateASIDWithPrefix generates a new ASID with custom prefix and vanity components
// Format: prefix(2) + random(2) + subject(5) + random(2) + predicate(7) + random(2) + context(5) + random(4) + actor(3)
// Returns an ID like "AS47ALICE22FOLLOWS33BOBXY5E7ACLA" or "JD47ALICE22FOLLOWS33BOBXY5E7ACLA"
// The prefix must be exactly 2 characters and will be uppercased
func GenerateASIDWithPrefix(prefix, subject, predicate, context, actor string) (string, error) {
	// Validate and normalize prefix
	prefix = strings.ToUpper(prefix)
	if len(prefix) != 2 {
		return "", fmt.Errorf("prefix must be exactly 2 characters, got %d", len(prefix))
	}

	// Generate a new UUID for the random components
	u, err := uuid.NewRandom()
	if err != nil {
		return "", err
	}

	// Extract vanity components with appropriate lengths
	subjectVanity := extractVanityComponent(subject, 5)
	predicateVanity := extractVanityComponent(predicate, 7)
	contextVanity := extractVanityComponent(context, 5)
	actorVanity := extractVanityComponent(actor, 3)

	// Get random hex characters from UUID (remove dashes)
	uuidStr := strings.ToUpper(strings.ReplaceAll(u.String(), "-", ""))

	// Extract random segments: 2 + 2 + 2 + 4 = 10 chars total
	random1 := uuidStr[0:2]  // First 2 chars
	random2 := uuidStr[2:4]  // Next 2 chars
	random3 := uuidStr[4:6]  // Next 2 chars
	random4 := uuidStr[6:10] // Next 4 chars

	// Construct ASID: prefix(2) + random(2) + subject(5) + random(2) + predicate(7) + random(2) + context(5) + random(4) + actor(3) = 32 chars
	asid := fmt.Sprintf("%s%s%s%s%s%s%s%s%s", prefix, random1, subjectVanity, random2, predicateVanity, random3, contextVanity, random4, actorVanity)

	return asid, nil
}

// GenerateASIDWithVanity generates a new ASID with vanity components from subject, predicate, context, and actor
// Format: AS + random(2) + subject(5) + random(2) + predicate(7) + random(2) + context(5) + random(4) + actor(3)
// Returns an ID like "AS47ALICE22FOLLOWS33BOBXY5E7ACLA"
func GenerateASIDWithVanity(subject, predicate, context, actor string) (string, error) {
	return GenerateASIDWithPrefix("AS", subject, predicate, context, actor)
}

// GenerateASID generates a new ASID with vanity components from subject, predicate, context, and actor
// Format: AS + random(2) + subject(5) + random(2) + predicate(7) + random(2) + context(5) + random(4) + actor(3)
// Returns an ID like "AS47ALICE22FOLLOWS33BOBXY5E7ACLA"
func GenerateASID(subject, predicate, context, actor string) (string, error) {
	return GenerateASIDWithVanity(subject, predicate, context, actor)
}

// GenerateASIDWithPrefixAndRetry generates an ASID with custom prefix, vanity components, and collision detection
// The checkExists function should return true if the ID already exists
func GenerateASIDWithPrefixAndRetry(prefix, subject, predicate, context, actor string, checkExists func(string) bool) (string, error) {
	const maxRetries = 10

	for i := 0; i < maxRetries; i++ {
		asid, err := GenerateASIDWithPrefix(prefix, subject, predicate, context, actor)
		if err != nil {
			return "", err
		}

		// Check if this ID already exists
		if !checkExists(asid) {
			return asid, nil
		}
	}

	// If we get here, we've had too many collisions
	// This is extremely unlikely with UUIDs, so just return the last attempt
	return GenerateASIDWithPrefix(prefix, subject, predicate, context, actor)
}

// GenerateASIDWithVanityAndRetry generates an ASID with vanity components and collision detection
// The checkExists function should return true if the ID already exists
func GenerateASIDWithVanityAndRetry(subject, predicate, context, actor string, checkExists func(string) bool) (string, error) {
	return GenerateASIDWithPrefixAndRetry("AS", subject, predicate, context, actor, checkExists)
}

// GenerateASIDWithRetry generates an ASID with collision detection
// The checkExists function should return true if the ID already exists
func GenerateASIDWithRetry(subject, predicate, context, actor string, checkExists func(string) bool) (string, error) {
	return GenerateASIDWithVanityAndRetry(subject, predicate, context, actor, checkExists)
}

// GenerateJobASID generates an async job ASID
// Format: JB + random(2) + jobType(5) + random(2) + process(7) + random(2) + source(5) + random(4) + actor(3)
// Returns an ID like "JB47JDING22PROCESS33URLXY5E7AGRU"
//
// The components provide semantic information:
//   - jobType: The type of job (jd_ingestion, candidate_scoring, batch_rescore)
//   - process: Fixed "process" predicate
//   - source: The source URL/file identifier
//   - actor: Who requested the job
//
// This format enables:
//   - Quick visual identification of job type
//   - Traceability to source and requester
//   - Consistency with ASID format throughout system
func GenerateJobASID(jobType, source, actor string) (string, error) {
	return GenerateASIDWithPrefix("JB", jobType, "process", source, actor)
}

// GenerateJobASIDWithRetry generates a Job ASID with collision detection
// The checkExists function should return true if the ID already exists
func GenerateJobASIDWithRetry(jobType, source, actor string, checkExists func(string) bool) (string, error) {
	return GenerateASIDWithPrefixAndRetry("JB", jobType, "process", source, actor, checkExists)
}

// GenerateExecutionID generates a Pulse Execution ID
// Format: PX (Pulse Execution) + timestamp + random
func GenerateExecutionID() string {
	return GenerateASIDSimple("PX", "execution", "pulse")
}

// GenerateASIDSimple generates a simple ASID without retry logic (for non-critical IDs)
func GenerateASIDSimple(prefix, subject, context string) string {
	id, _ := GenerateASIDWithPrefix(prefix, subject, "id", context, "")
	return id
}

// IsValidASID checks if a string is a valid ASID format
// Supports any 2-letter prefix (AS, JD, JB, etc.), legacy format (prefix + 30 hex chars),
// and new vanity format with mixed segments
func IsValidASID(id string) bool {
	// Must be 32 characters total
	if len(id) != 32 {
		return false
	}

	// Prefix must be 2 uppercase letters or digits
	for i := 0; i < 2; i++ {
		r := rune(id[i])
		if !((r >= 'A' && r <= 'Z') || (r >= '0' && r <= '9')) {
			return false
		}
	}

	// Check if it's the legacy format (all remaining chars are hex)
	isLegacyFormat := true
	for _, r := range id[2:] {
		if !((r >= '0' && r <= '9') || (r >= 'A' && r <= 'F')) {
			isLegacyFormat = false
			break
		}
	}

	if isLegacyFormat {
		return true
	}

	// If not legacy format, check if it's valid new vanity format
	// Format: prefix(2) + random(2) + subject(5) + random(2) + predicate(7) + random(2) + context(5) + random(4) + actor(3)

	// Check random segments (positions: 2-3, 9-10, 18-19, 25-28) - must be hex
	randomSegments := [][]int{{2, 4}, {9, 11}, {18, 20}, {25, 29}}
	for _, segment := range randomSegments {
		for i := segment[0]; i < segment[1]; i++ {
			r := rune(id[i])
			if !((r >= '0' && r <= '9') || (r >= 'A' && r <= 'F')) {
				return false
			}
		}
	}

	// Check vanity segments (positions: 4-8, 11-17, 20-24, 29-31) - must be alphanumeric
	vanitySegments := [][]int{{4, 9}, {11, 18}, {20, 25}, {29, 32}}
	for _, segment := range vanitySegments {
		for i := segment[0]; i < segment[1]; i++ {
			r := rune(id[i])
			if !((r >= '0' && r <= '9') || (r >= 'A' && r <= 'Z')) {
				return false
			}
		}
	}

	return true
}

// IsVanityASID checks if an ASID uses the new vanity format
func IsVanityASID(id string) bool {
	if !IsValidASID(id) {
		return false
	}

	// Check if it's the legacy format (all remaining chars are hex)
	for _, r := range id[2:] {
		if !((r >= '0' && r <= '9') || (r >= 'A' && r <= 'F')) {
			return true // Contains non-hex, so it's vanity format
		}
	}

	return false // All chars are hex, so it's legacy format
}

File: vanity-id-rs/src/asid.rs
//! Application-Scoped ID (ASID) generation functionality.
//!
//! This module provides functions for generating structured 32-character identifiers
//! with semantic vanity components mixed with random segments for uniqueness.
//!
//! ASID Format: prefix(2) + random(2) + subject(5) + random(2) + predicate(7) + random(2) + context(5) + random(4) + actor(3)
//! Example: "AS47ACMEX22ENGINEE33SFBAY5E7AJOB"

use sha2::{Digest, Sha256};
use uuid::Uuid;

use crate::id::IdError;

/// Generates a Job Description ASID
/// Format: JD + random(2) + company(5) + random(2) + role(7) + random(2) + location(5) + random(4) + type(3)
/// Returns an ID like "JD47ACMEX22ENGINEE33SFBAY5E7AJOB"
pub fn generate_jd_asid(company: &str, role_title: &str, location: &str) -> Result<String, IdError> {
    let u = Uuid::new_v4();
    let uuid_str: String = u.to_string().replace('-', "").to_uppercase();

    // Extract random segments: 2 + 2 + 2 + 4 = 10 chars total
    let rand1 = &uuid_str[0..2];
    let rand2 = &uuid_str[2..4];
    let rand3 = &uuid_str[4..6];
    let rand4 = &uuid_str[6..10];

    // Extract vanity components
    let company_part = extract_vanity_component(company, 5);
    let role_part = extract_vanity_component(role_title, 7);
    let location_part = extract_vanity_component(location, 5);
    let type_part = "JOB"; // Default type indicator for job descriptions

    // Construct the JD ASID - 32 characters total
    Ok(format!(
        "JD{}{}{}{}{}{}{}{}",
        rand1, company_part, rand2, role_part, rand3, location_part, rand4, type_part
    ))
}

/// Generates a JD ASID with collision detection
pub fn generate_jd_asid_with_retry<F>(
    company: &str,
    role_title: &str,
    location: &str,
    check_exists: F,
) -> Result<String, IdError>
where
    F: Fn(&str) -> bool,
{
    const MAX_RETRIES: usize = 10;

    for _ in 0..MAX_RETRIES {
        let jd_id = generate_jd_asid(company, role_title, location)?;

        if !check_exists(&jd_id) {
            return Ok(jd_id);
        }
        // Collision detected, retry
    }

    Err(IdError::GenerationFailed(format!(
        "failed to generate unique JD ASID after {} attempts",
        MAX_RETRIES
    )))
}

/// Extracts the specified number of characters from a string, using hash-based padding
fn extract_vanity_component(s: &str, length: usize) -> String {
    if s.is_empty() {
        return generate_hash_based_padding("", length);
    }

    // Clean and uppercase the string, keeping only alphanumeric characters
    let cleaned: String = s
        .to_uppercase()
        .chars()
        .filter(|c| c.is_ascii_alphanumeric())
        .collect();

    if cleaned.is_empty() {
        return generate_hash_based_padding(s, length);
    }

    // Take first N characters, pad with hash-based characters if needed
    if cleaned.len() >= length {
        return cleaned[..length].to_string();
    }

    // Need padding - use hash-based approach
    let needed = length - cleaned.len();
    let padding = generate_hash_based_padding(s, needed);
    format!("{}{}", cleaned, padding)
}

/// Generates deterministic padding characters from string hash
fn generate_hash_based_padding(s: &str, length: usize) -> String {
    if length == 0 {
        return String::new();
    }

    // Valid characters for vanity components (alphanumeric, excluding confusing ones)
    const VALID_CHARS: &str = "ABCDEFGHJKLMNPQRSTUVWXYZ23456789"; // No I,O,0,1 for clarity

    let mut hasher = Sha256::new();
    hasher.update(s.as_bytes());
    let hash = hasher.finalize();

    let valid_chars: Vec<char> = VALID_CHARS.chars().collect();
    let mut result = String::with_capacity(length);

    for i in 0..length {
        let byte_index = i % hash.len();
        let char_index = (hash[byte_index] as usize) % valid_chars.len();
        result.push(valid_chars[char_index]);
    }

    result
}

/// Generates a new ASID with custom prefix and vanity components
/// Format: prefix(2) + random(2) + subject(5) + random(2) + predicate(7) + random(2) + context(5) + random(4) + actor(3)
pub fn generate_asid_with_prefix(
    prefix: &str,
    subject: &str,
    predicate: &str,
    context: &str,
    actor: &str,
) -> Result<String, IdError> {
    // Validate and normalize prefix
    let prefix = prefix.to_uppercase();
    if prefix.len() != 2 {
        return Err(IdError::InvalidLength(format!(
            "prefix must be exactly 2 characters, got {}",
            prefix.len()
        )));
    }

    let u = Uuid::new_v4();
    let uuid_str: String = u.to_string().replace('-', "").to_uppercase();

    // Extract vanity components with appropriate lengths
    let subject_vanity = extract_vanity_component(subject, 5);
    let predicate_vanity = extract_vanity_component(predicate, 7);
    let context_vanity = extract_vanity_component(context, 5);
    let actor_vanity = extract_vanity_component(actor, 3);

    // Extract random segments: 2 + 2 + 2 + 4 = 10 chars total
    let random1 = &uuid_str[0..2];
    let random2 = &uuid_str[2..4];
    let random3 = &uuid_str[4..6];
    let random4 = &uuid_str[6..10];

    // Construct ASID: 32 chars total
    Ok(format!(
        "{}{}{}{}{}{}{}{}{}",
        prefix,
        random1,
        subject_vanity,
        random2,
        predicate_vanity,
        random3,
        context_vanity,
        random4,
        actor_vanity
    ))
}

/// Generates a new ASID with vanity components (AS prefix)
pub fn generate_asid_with_vanity(
    subject: &str,
    predicate: &str,
    context: &str,
    actor: &str,
) -> Result<String, IdError> {
    generate_asid_with_prefix("AS", subject, predicate, context, actor)
}

/// Generates a new ASID with vanity components (AS prefix)
pub fn generate_asid(
    subject: &str,
    predicate: &str,
    context: &str,
    actor: &str,
) -> Result<String, IdError> {
    generate_asid_with_vanity(subject, predicate, context, actor)
}

/// Generates an ASID with custom prefix, vanity components, and collision detection
pub fn generate_asid_with_prefix_and_retry<F>(
    prefix: &str,
    subject: &str,
    predicate: &str,
    context: &str,
    actor: &str,
    check_exists: F,
) -> Result<String, IdError>
where
    F: Fn(&str) -> bool,
{
    const MAX_RETRIES: usize = 10;

    for _ in 0..MAX_RETRIES {
        let asid = generate_asid_with_prefix(prefix, subject, predicate, context, actor)?;

        if !check_exists(&asid) {
            return Ok(asid);
        }
    }

    // If we get here, we've had too many collisions
    // This is extremely unlikely with UUIDs, so just return the last attempt
    generate_asid_with_prefix(prefix, subject, predicate, context, actor)
}

/// Generates an ASID with vanity components and collision detection
pub fn generate_asid_with_vanity_and_retry<F>(
    subject: &str,
    predicate: &str,
    context: &str,
    actor: &str,
    check_exists: F,
) -> Result<String, IdError>
where
    F: Fn(&str) -> bool,
{
    generate_asid_with_prefix_and_retry("AS", subject, predicate, context, actor, check_exists)
}

/// Generates an ASID with collision detection
pub fn generate_asid_with_retry<F>(
    subject: &str,
    predicate: &str,
    context: &str,
    actor: &str,
    check_exists: F,
) -> Result<String, IdError>
where
    F: Fn(&str) -> bool,
{
    generate_asid_with_vanity_and_retry(subject, predicate, context, actor, check_exists)
}

/// Generates an async job ASID
/// Format: JB + random(2) + jobType(5) + random(2) + process(7) + random(2) + source(5) + random(4) + actor(3)
pub fn generate_job_asid(job_type: &str, source: &str, actor: &str) -> Result<String, IdError> {
    generate_asid_with_prefix("JB", job_type, "process", source, actor)
}

/// Generates a Job ASID with collision detection
pub fn generate_job_asid_with_retry<F>(
    job_type: &str,
    source: &str,
    actor: &str,
    check_exists: F,
) -> Result<String, IdError>
where
    F: Fn(&str) -> bool,
{
    generate_asid_with_prefix_and_retry("JB", job_type, "process", source, actor, check_exists)
}

/// Generates a Pulse Execution ID
pub fn generate_execution_id() -> String {
    generate_asid_simple("PX", "execution", "pulse")
}

/// Generates a simple ASID without retry logic (for non-critical IDs)
pub fn generate_asid_simple(prefix: &str, subject: &str, context: &str) -> String {
    generate_asid_with_prefix(prefix, subject, "id", context, "")
        .unwrap_or_else(|_| format!("{}00000000000000000000000000000", prefix))
}

/// Checks if a string is a valid ASID format
/// Supports any 2-letter prefix, legacy format, and new vanity format
pub fn is_valid_asid(id: &str) -> bool {
    // Must be 32 characters total
    if id.len() != 32 {
        return false;
    }

    let chars: Vec<char> = id.chars().collect();

    // Prefix must be 2 uppercase letters or digits
    for i in 0..2 {
        let c = chars[i];
        if !c.is_ascii_uppercase() && !c.is_ascii_digit() {
            return false;
        }
    }

    // Check if it's the legacy format (all remaining chars are hex)
    let is_legacy_format = id[2..].chars().all(|c| c.is_ascii_hexdigit());

    if is_legacy_format {
        return true;
    }

    // If not legacy format, check if it's valid new vanity format
    // Check random segments (positions: 2-3, 9-10, 18-19, 25-28) - must be hex
    let random_segments = [(2, 4), (9, 11), (18, 20), (25, 29)];
    for (start, end) in random_segments {
        for i in start..end {
            let c = chars[i];
            if !c.is_ascii_hexdigit() {
                return false;
            }
        }
    }

    // Check vanity segments (positions: 4-8, 11-17, 20-24, 29-31) - must be alphanumeric
    let vanity_segments = [(4, 9), (11, 18), (20, 25), (29, 32)];
    for (start, end) in vanity_segments {
        for i in start..end {
            let c = chars[i];
            if !c.is_ascii_alphanumeric() {
                return false;
            }
        }
    }

    true
}

/// Checks if an ASID uses the new vanity format
pub fn is_vanity_asid(id: &str) -> bool {
    if !is_valid_asid(id) {
        return false;
    }

    // Check if it's the legacy format (all remaining chars are hex)
    !id[2..].chars().all(|c| c.is_ascii_hexdigit())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_generate_jd_asid() {
        let id = generate_jd_asid("Acme Corp", "Software Engineer", "San Francisco").unwrap();
        assert_eq!(id.len(), 32);
        assert!(id.starts_with("JD"));
    }

    #[test]
    fn test_generate_asid() {
        let id = generate_asid("Alice", "follows", "Twitter", "User").unwrap();
        assert_eq!(id.len(), 32);
        assert!(id.starts_with("AS"));
    }

    #[test]
    fn test_generate_asid_with_prefix() {
        let id = generate_asid_with_prefix("JB", "import", "process", "url", "sys").unwrap();
        assert_eq!(id.len(), 32);
        assert!(id.starts_with("JB"));
    }

    #[test]
    fn test_is_valid_asid() {
        let id = generate_asid("test", "test", "test", "tst").unwrap();
        assert!(is_valid_asid(&id));
        assert!(!is_valid_asid("short"));
        assert!(!is_valid_asid(""));
    }

    #[test]
    fn test_is_vanity_asid() {
        let id = generate_asid("Alice", "follows", "Bob", "User").unwrap();
        assert!(is_vanity_asid(&id));
    }

    #[test]
    fn test_generate_execution_id() {
        let id = generate_execution_id();
        assert_eq!(id.len(), 32);
        assert!(id.starts_with("PX"));
    }

    #[test]
    fn test_extract_vanity_component_padding() {
        // Short input should be padded deterministically
        let comp1 = extract_vanity_component("AB", 5);
        let comp2 = extract_vanity_component("AB", 5);
        assert_eq!(comp1.len(), 5);
        assert_eq!(comp1, comp2); // Should be deterministic
    }
}
